File: c/202_Happy-Number.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

int divde_square(int n){
    int s = 0;
    int mod;
    while(n>0){
        mod = n%10;
        n=n/10;
        s += (mod*mod);
    }
    return s;
}

bool isHappy(int n){
    
    int hashmap[731] = {};
    
    if(n==1)
        return true;
    while(true){
        n = divde_square(n);
        if(n==1)
            return true;
        if(hashmap[n]==-1)
            return false;
        hashmap[n]=-1;
    }
}


int main(){
    int num = 7;
    printf("%d", isHappy(num));
    return 0;
}

File: python/202_Happy-Number.py
class Solution:
    def isHappy(self, n: int) -> bool:
        def divde_square(x):
            s=0
            while x>0:
                x, mod = divmod(x,10)
                s += (mod**2)
            return s
        
        if n==1:
            return True

        _set = set()
        _set.add(n)
        while True:
            n = divde_square(n)
            if n == 1:
                return True
            if n in _set:
                return False
            _set.add(n)


def main():
    sol = Solution()
    n = 19
    print(sol.isHappy(n))
    n=2
    print(sol.isHappy(n))

if __name__ == '__main__':
    main()
